enhancements/authentication/pod-security-admission-enforcement.md

```diff
@@ -0,0 +1,315 @@
+type PSAEnforcementConfigSpec struct {
+	// enforcementMode gives the user different options:
+	// - Restricted enables the cluster to move to PSA enforcement, if there are no violating Namespaces detected.
+	//   If violating Namespaces are found, the operator moves into "Upgradeable=false".
+	// - Privileged enables the cluster to opt-out from PSA enforcement for now and it resolves the operator status of "Upgradeable=false" in case of violating Namespaces.
+	//
+	// defaults to "Restricted"
+	EnforcementMode PSAEnforcementMode `json:"enforcementMode"`
+}
+
+// PSAEnforcementConfigStatus is a struct that signals to the user, if the cluter is going to start with PSA enforcement and if there are any violating Namespaces.
+type PSAEnforcementConfigStatus struct {
+	// enforcementMode indicates if PSA enforcement will happen:
+	// - "Restricted" indicates that enforcement is possible and will happen.
+	// - "Privileged" indidcates that either enforcement will not happen:
+	//   - either it is not wished or
+	//   - it isn't possible without potentially breaking workloads.
+	EnforcementMode PSAEnforcementMode `json:"enforcementMode"`
+
+	// violatingNamespaces lists Namespaces that are violating. Needs to be resolved in order to move to Restricted.
+	//
+	// +optional
+	ViolatingNamespaces []ViolatingNamespace `json:"violatingNamespaces,omitempty"`
+}
+
+// ViolatingNamespace provides information about a namespace that cannot comply
+// with the chosen enforcement mode.
+type ViolatingNamespace struct {
+	// name is the Namespace that has been flagged as potentially violating if
+	// enforced.
+	Name string `json:"name"`
+
+	// reason is a textual description explaining why the Namespace is incompatible
+	// with the requested Pod Security mode and highlights which mode is affected.
+	//
+	// Possible values are:
+	// - PSAConfig: Misconfigured OpenShift Namespace
+	// - PSAConfig: PSA label syncer disabled
+	// - PSALabel: ServiceAccount with insufficient SCCs
+	//
+	// +optional
+	Reason string `json:"reason,omitempty"`
+}
+```
+
+Here is a boolean table with the expected outcomes:
+
+| `spec.enforcementMode` | length of `status.violatingNamespaces` | `status.enforcementMode` | `OperatorStatus`  |
+| ---------------------- | -------------------------------------- | ------------------------ | ----------------- |
+| Privileged             | More than 0                            | Privileged               | AsExpected        |
+| Privileged             | 0                                      | Privileged               | AsExpected        |
+| Restricted             | More than 0                            | Privileged               | Upgradeable=False |
+| Restricted             | 0                                      | Restricted               | AsExpected        |
+
+If a user encounters `status.violatingNamespaces` it is expected to:
+
+- resolve the violating Namespaces in order to be able to `Upgrade` or
```

Comment:
A workflow description of how to resolve violating namespaces would be really useful. What are the options for getting a namespace out of violating

Reply:
https://github.com/openshift/enhancements/pull/1747/files#diff-e48dd33b3f59b545b505293e5c81d96a29a2e9cc994f4527b46bde63a2def9d8R354-R398